lib/associations/associations.rb

```diff
@@ -92,65 +103,72 @@ def self.included(base)
     end
 
     module Methods
-      def has_many(association_id, options = {})
+      def association_metadata(type, association_name, options = {})
+        association_names = association_name.to_s
+        association_class = (options[:class_name] || association_names.classify).constantize
+        {
+          target_model: association_class,
+          primary_key: normalize_primary_key(type, association_class),
+          foreign_key: normalize_foreign_key(type, association_name, options)
+        }.merge(options)
+      end
 
-        define_method(association_id) do
-          options = {
-            :class_name => association_id.to_s.classify,
-            :foreign_key => self.class.to_s.foreign_key,
-            :primary_key => self.class.primary_key
-          }.merge(options)
-
-          klass = options[:class_name].constantize
-          primary_key_value = send(options[:primary_key])
-          foreign_key = options[:foreign_key].to_sym
-
-          if Object.const_defined?(:ActiveRecord) && ActiveRecord.const_defined?(:Relation) && klass < ActiveRecord::Relation
-            klass.where(foreign_key => primary_key_value)
-          elsif klass.respond_to?(:scoped)
-            klass.scoped(:conditions => {foreign_key => primary_key_value})
+      def normalize_primary_key(type, association_class)
+        primary_key = type == :belongs_to ? association_class.primary_key : name.constantize.primary_key
+        primary_key.to_sym
+      end
+
+      def normalize_foreign_key(type, association_name, options)
+        foreign_key =
+          if type == :belongs_to
+            options[:foreign_key] || (options[:class_name] || association_name).to_s.foreign_key
           else
-            klass.where(foreign_key => primary_key_value)
+            name.foreign_key
           end
-        end
+        foreign_key.to_sym
```

Comment:
@tiagopog similar to https://github.com/zilkey/active_hash/pull/158/files#r161891497